Add support for RTL82328 dialect

mrnuke opened the pull request:
Add initial support for the PoE dialect used by RTL8238B PSE chips.

The first attempt at an API consisted of a struct of function pointers
representing a set of actions. A dialect would have to implement its
own set of handlers by populating these pointers. Common code could be
placed in shared functions and reused across dialects. This approach
did not allow the PoE reply handlers to be shared. Each dialect needed
to maintain a mapping of command IDs to reply_handler function.

The second attempt at an API defined a new set of actions as an enum,
and expected each dialect to provide a mapping from realtek-poe
actions to command IDs. This works well because the command IDs are
the main difference between the realtek and broadcom protocols. The
majority of existing code can be reused. The mapping can be reversed
in order to translate a command id to a generic packet handler. This
solves the per-dialect duplication of the reply handler.

However, the dialects do not map 1:1. The structure of some commands
in the realtek dialect is noticeably different from equivalent
commands in the broadcom dialect.

This creates the third and current mapping API. Use the mapping API as
the main way to differentiate dialects. For incompatible commands,
introduce introduce a structure of function pointers to allow per-
dialect customizations. This approach, hybrid as it may be, solves
the issues encountered with the previous two attempts.

```
struct dialect_ops {
	int (*init_async)(struct mcu *mcu, const struct config *cfg);
	int (*poll_async)(struct mcu *mcu, const struct config *cfg);
	int (*reset)(struct mcu *mcu);
	int (*handle_reply)(struct mcu_state *mcu, uint8_t *reply, size_t len);
};

struct dialect_map {
	const struct dialect_map_entry *entries;
	size_t len;
};

struct dialect_desc {
	const struct dialect_ops *ops;
	const struct dialect_map *map;
};
```

Many commands are very similar between the the two dialects. The command IDs differ. The approach is to have common code implementing the dialects, and have a mapping bewteen command and command_id of the dialect:


```
enum poe_cmd {
	POE_CMD_NONE = 0,
	POE_CMD_SET_XYZ,
	...
	POE_CMD_SET_YZW,
	...
};

poe_cmd_set_xyz(struct poe_dialect *dialect, ...)
{
	uint8_t cmd = dialect_map(dialect, POE_CMD_SET_XYZ);

	poe_cmd_queue(cmd, ...);
}

const struct dialect_map __map = {
	[POE_CMD_SET_XYZ] = { 0x42, CMD_4_PORT, CMD_ALL_PORT },
	...
};
```

The flags may seem obscure. Most command dealing with a port number can take more than one (port, value) argument pair. For example `01 <val>02 <val> 03 <val> ff ff`, up to four pairs per command, with `ff` meaning "nothing to see here. This is the use case for the 4port command function:

```
static int poet_cmd_4_port(uint8_t cmd_id, uint8_t port[4], uint8_t data[4])
{
	uint8_t cmd[] = { cmd_id, 0x00, port[0], data[0], port[1], data[1],
		port[2], data[2], port[3], data[3] };

		return poe_cmd_queue(cmd, sizeof(cmd));
}
```

Other commands can use `7f` as the port ID to configure all ports with the same value: `7f <val>`. The flags in `dialect` map would indicate which commands can use such optimizations.

I am hoping that restructuring the code in this manner would reduce duplication, and actually make a PR that makes some sense.

mrnuke committed:
dialect_rtl: add RTL8238B compatible dialect

Implement the realtek dialect found on RTL8238B PSE controllers using
the mapping API. MOst of the commands are identical to the broadcom
dialect, except for the command IDs.

the port status packet does not take a list of ports, but instead
starts form a port id (divided by 4). Account for this difference
using the poll_async() method of the dialect struct.

PSE to port mapping is also in a different byte of the ext config
packet. However, most of these differences are easy to account for.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke committed:
realtek-poe: automatically reset mcu on timeouts

With the realtek dialect, an unresponsive MCU may be revived by
sending a reset command. Implement a timeout mechanism that send a
reset command after a generous wait period -- currently 2 seconds.

Since the broadcom dialect does not implement a .reset method, no
change in behavior is expected there.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke committed:
realtek-poe: define API to map multiple dialects

realtek-poe supports the protocol originally found on Broadcom chips,
and now referred to as the "broadcom dialect". A similar protocol
appeared from Realtek on their RTL8238B PoE controllers. The latter
is referred to as the "realtek dialect". These are similar enough to
be considered dialects of the same protocol, rather than completely
separate protocols.

The first attempt at an API consisted of a struct of function pointers
representing a set of actions. A dialect would have to implement its
own set of handlers by populating these pointers. Common code could be
placed in shared functions and reused across dialects. This approach
did not allow the PoE reply handlers to be shared. Each dialect needed
to maintain a mapping of command IDs to reply_handler function.

The second attempt at an API defined a new set of actions as an enum,
and expected each dialect to provide a mapping from realtek-poe
actions to command IDs. This works well because the command IDs are
the main difference between the realtek and broadcom protocols. The
majority of existing code can be reused. The mapping can be reversed
in order to translate a command id to a generic packet handler. This
solves the per-dialect duplication of the reply handler.

However, the dialects do not map 1:1. The structure of some commands
in the realtek dialect is noticeably different from equivalent
commands in the broadcom dialect.

This creates the third and current mapping API. Use the mapping API as
the main way to differentiate dialects. For incompatible commands,
introduce introduce a structure of function pointers to allow per-
dialect customizations. This approach, hybrid as it may be, solves
the issues encountered with the previous two attempts.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke committed:
realtek-poe: rewrite broadcom dialect in terms of the mapping API

Instead of using hardcoded command IDs, use realtek-poe command enums.
Use a 1:1 mapping between the enums and broadcom dialect command IDs
in order to place the correct ID in the packet. To find the correct
reply handler, use the reverse map.

The forward and reverse lookups are both O(1) in time, and thus add
negligible overhead. Use this as an example of how a new dialect could
be implemented.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke committed:
tek-poe: allow changing baudrate dialect in the config

A smart program should automatically detect the dialect, including the
baudrate. A program that requires the user to correctly configure
things may be easier to write, but is harder to use.

Ideally, we'd want to send a set of "safe" commands, and based on the
responses, figure out the correct dialect to use. This also applies to
finding out the correct baudrate. I've been meaning to implement this
for a number of years, and I have had neither inspiration, not time to
do so.

Instead of delaying realtek dialect support, add temporary code that
allows the user to override the dialect settings. While not ideal,
this allows realtek-poe to be deployed to, and useful to more users.
It seems better than keeping the realtek dialect support contained to
an obscure branch.

For anyone looking to implement the missing dialect detection, a good
resource for autodetecting serial protocols is gpsd.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

Neustradamus commented:
To follow this PR :)

mrnuke committed:
dialect: realtek: Change mystery parameter from 1 to 0

Command 0x08 affects byte [04] of the get_port_config_reply. This is
documented by svanheule to mean "auto_powerup", with no additional
explanation.

If we set this to 1, like we used to, "realtek" dialect devices get
stuck to "Requesting power". When set to 0 instead, the state goes all
the way to "Delivering power. Set this parameter to zero, and call it
a "mystery" parameter.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke committed:
realtek-poe: handle 4-port replies

Commands sent by poet_setup() control parameters for up to 4 ports,
hence the 4-port name. These commands do send replies, indicating the
status for each port. They can indicate if something went wrong during
configuration, so process these replies and log any failures.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

plappermaul commented:
Just for completeness ...

https://github.com/plappermaul/realtek-doc/blob/main/RTL8238B%20Datasheet.pdf

https://github.com/plappermaul/realtek-doc/blob/main/RTL8238B%20Host%20Command%20Guide.pdf

mrnuke committed:
realtek-poe: use the number of detected ports to check port IDs

When parsing the port status reply, realtek-poe relied on the MAX_PORT
constant to validate a port ID. However, at that point, we know how
many ports the PoE controller has. That comes via num_detected_ports,
which should be populated at as part of the initialization sequence.

Use state->num_detected_ports as the upper limit for a port ID instead
of MAX_PORT.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>

mrnuke commented:
@Hurricos , I am ready for you to begin your review. The main blocker has been my wish to implement dialect auto-detection. However, it seems that that is a classic case of letting perfect be the enemy of good. I don't want to delay this any longer, as the realtek dialect code otherwise works okay.

There are a few changes I want to make to the data structures, but I feel that the mapping API and split between ops and raw mapping is a reliable solution at this point.

plappermaul commented on src/main.c:
```diff
@@ -358,6 +358,17 @@ static int poe_cmd_port_disconnect_type(struct mcu *mcu, uint8_t port,
+ *	reasons RTL8238B ports get stuck on "Requesting power" status, and
+ *	never transition to "Delivering power".
+ */
+static int poe_cmd_mystery_parameter(struct mcu *mcu, uint8_t ports[4],
```
Thank you for enhancing the realtek-poe package. Why would we call this a mystery parameter if we have all documentation at hand? https://github.com/plappermaul/realtek-doc/blob/main/datasheets/RTL8238B%20Host%20Command%20Guide.pdf

mrnuke replied:
Because the documentation wasn't available when I wrote the code, and i like to leave easter eggs here and there.

aparcar replied:
I'm not entirely sure what's the best way forward here but I just got myself a 24EP and really need proper PoE support, could we solve this and merge this please :)?

plappermaul replied:
I have no issues with this. This was just a reminder that we have mor knowledge now. Merging this PR is up to @Hurricos and @mrnuke

Hurricos commented:
Sorry for the lack of warning.

Sounds like there is consensus to merge here?

Hurricos commented:
Hmm, I'm not going to remember to merge this otherwise.

Chaos! It's merged! Hahahahahaha.

What I don't know off the top of my head is what's needed upstream. I believe we need a new package bump in openwrt/packages. @mrnuke or @aparcar or @plappermaul - could you spin up a PR there?

Klaas- commented:
@Hurricos I guess you need a tagged version that includes this change and then we can do a PR towards openwrt packages like https://github.com/openwrt/packages/pull/26854/files

mrnuke commented:
I pushed the v1.3 tag. I think the rest is as simple as changing the `PKG_VERSION` in the [packages feed](https://github.com/openwrt/packages/blob/7be94fac3708ba2ae15abcb2a92435a57b05dd37/utils/realtek-poe/Makefile#L6C1-L6C12).

hailfinger commented:
Something like this? https://github.com/openwrt/packages/pull/27386
